packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1137,127 +1137,159 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
-        continue;
       }
     }
 
-    finalResolution.addAll(
-      pluginResolution.map((Plugin plugin) =>
+    pluginResolutions.addAll(
+      pluginResolution.values.map((Plugin plugin) =>
           PluginInterfaceResolution(plugin: plugin, platform: platformKey)),
     );
   }
   if (hasResolutionError) {
     throwToolExit('Please resolve the plugin implementation selection errors');
   }
-  return finalResolution;
+  return pluginResolutions;
+}
+
+/// Get the [resolutionPluginName] as the first choice for the plugin
+/// implementation and the [defaultImplementation] as fallback.
+(String? resolutionPluginName, String? defaultImplementation) _getPluginImplementationCandidate(Plugin plugin, String platformKey) {
+  String? defaultImplementation = plugin.defaultPackagePlatforms[platformKey];
+  if (plugin.platforms[platformKey] == null && defaultImplementation == null) {
+    // The plugin doesn't implement this platform.
+    return (null, null);
+  }
+  String? implementsPackage = plugin.implementsPackage;
+  final bool hasInlineDartImplementation =
+      plugin.pluginDartClassPlatforms[platformKey] != null &&
+          plugin.pluginDartClassPlatforms[platformKey] != 'none';
+  if (implementsPackage == null || implementsPackage.isEmpty) {
+    if (defaultImplementation == null && !hasInlineDartImplementation) {
+      // Skip native inline PluginPlatform implementation
+      return (null, null);
+    }
+    if (defaultImplementation != null) {
+      return (null, defaultImplementation);
+    } else {
+      // An app-facing package (i.e., one with no 'implements') with an
+      // inline implementation should be its own default implementation.
+      // Desktop platforms originally did not work that way, and enabling
+      // it unconditionally would break existing published plugins, so
+      // only treat it as such if either:
+      // - the platform is not desktop, or
+      // - the plugin requires at least Flutter 2.11 (when this opt-in logic
+      //   was added), so that existing plugins continue to work.
+      // See https://github.com/flutter/flutter/issues/87862 for details.
+      final bool isDesktop = platformKey == 'linux' || platformKey == 'macos' || platformKey == 'windows';
+      final semver.VersionConstraint? flutterConstraint = plugin.flutterConstraint;
+      final semver.Version? minFlutterVersion = flutterConstraint != null &&
+          flutterConstraint is semver.VersionRange ? flutterConstraint.min : null;
+      final bool hasMinVersionForImplementsRequirement = minFlutterVersion != null &&
+          minFlutterVersion.compareTo(semver.Version(2, 11, 0)) >= 0;
+      if (!isDesktop || hasMinVersionForImplementsRequirement) {
+        // Given: implementsPackage == null, defaultImplementation == null, hasInlineDartImplementation == true
+        implementsPackage = plugin.name;
+        defaultImplementation = plugin.name;
+        return (implementsPackage, defaultImplementation);
```

Comment:
Related to my confusion expressed above: this does not appear to match the description of what the method returns. If an app-facing package is its own *default* implementation, that does not mean that it's the preferred implementation. An app-facing package's internal default should still be overridable by an explicit dependency on an alternate implementation.

Reply:
Thank you very much for the fast and helpful feedback. The method description was admittedly misleading.
I know added a precise method description and renamed some variables to gain more clarity in 12a38f9921d0390979945fcfe7e04f03842f2246.
If anything seems odd to you, let me know!